File: 0075_Sort_Colors/solution.go
/*
75. Sort Colors

Level: Medium

https://leetcode.com/problems/sort-colors
*/

package leetcode0075

/*

 */
func sortColors(nums []int) {
	i, j, k := 0, 0, len(nums)-1

	// for 循环中， nums[i:j] 中始终全是 1
	// 循环结束后，
	// nums[:i] 中全是 0
	// nums[j:] 中全是 2
	for j <= k {
		switch nums[j] {
		case 0:
			nums[i], nums[j] = nums[j], nums[i]
			i++
			j++
		case 1:
			j++
		case 2:
			nums[j], nums[k] = nums[k], nums[j]
			k--
		}
	}
}

File: 0075_Sort_Colors/solution.js
/*
75. Sort Colors

Level: Medium

https://leetcode.com/problems/sort-colors
*/

/*
  Solution: 
*/

/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function (nums) {
  if (nums == null || nums.length == 0) {
    return;
  }
  let lastRed = 0,
    firstBlue = nums.length - 1;
  let i = 0;
  while (i <= firstBlue) {
    if (nums[i] == 0 && i != lastRed) {
      swap(nums, lastRed++, i);
    } else if (nums[i] == 2 && i != firstBlue) {
      swap(nums, i, firstBlue--);
    } else ++i;
  }
};
const swap = (nums, i, j) => {
  const temp = nums[i];
  nums[i] = nums[j];
  nums[j] = temp;
}
export { sortColors };
